pkg/rules/terraswap/col4_logfinders_test.go

```diff
@@ -0,0 +1,181 @@
+package terraswap
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	"github.com/dezswap/cosmwasm-etl/pkg/eventlog"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_CreateCreateLogFinder(t *testing.T) {
+	var logFinder eventlog.LogFinder
+	var eventLogs eventlog.LogResults
+	var err error
+	setUp := func(factoryAddr, rawLogsStr string) {
+		logFinder = nil
+		eventLogs = eventlog.LogResults{}
+		logFinder, err = CreateCreatePairRuleFinder(factoryAddr)
+		if err != nil {
+			panic(err)
+		}
+		if err := json.Unmarshal([]byte(rawLogsStr), &eventLogs); err != nil {
+			panic(err)
+		}
+	}
+
+	tcs := []struct {
+		factoryAddr       string
+		rawLogStr         string
+		expectedResultLen int
+		errMsg            string
+	}{
+		{FactoryAddress[ClassicV1Key], CreatePairRawLogStr, 1, "must match once"},
+		{FactoryAddress[ClassicV1Key], createTwiceLogStr, 2, "must match twice"},
+		{FactoryAddress[ClassicV1Key], differentTypeLogsStr, 0, "must not match with different type"},
+		{FactoryAddress[ClassicV1Key], "[]", 0, "must not match with empty logs"},
+	}
+
+	for idx, tc := range tcs {
+		errMsg := fmt.Sprintf("idx(%d): %s", idx, tc.errMsg)
+		assert := assert.New(t)
+
+		setUp(tc.factoryAddr, tc.rawLogStr)
+		matchedResults := logFinder.FindFromLogs(eventLogs)
+		assert.Len(matchedResults, tc.expectedResultLen, errMsg)
+		if tc.expectedResultLen > 0 {
+			assert.Len(matchedResults[0], CreatePairMatchedLen, "must return all matched value")
+		}
+	}
+}
+
+func Test_LogFinders(t *testing.T) {
+	var logFinder eventlog.LogFinder
+	var eventLogs eventlog.LogResults
+
+	setUp := func(rawLogsStr string, pairs map[string]bool, finderFunc func(map[string]bool) (eventlog.LogFinder, error)) {
+		var err error
+		logFinder = nil
+		eventLogs = eventlog.LogResults{}
+		logFinder, err = finderFunc(pairs)
+		if err != nil {
+			panic(err)
+		}
+		if err := json.Unmarshal([]byte(rawLogsStr), &eventLogs); err != nil {
+			panic(err)
+		}
+	}
+
+	tcs := []struct {
+		rawLogStr         string
+		pairs             map[string]bool
+		finderFunc        func(map[string]bool) (eventlog.LogFinder, error)
+		expectedResultLen int
+		matchedLen        int
+		errMsg            string
+	}{
+		//Swap
+		{PairSwapRawLogStr, nil, CreatePairCommonRulesFinder, 1, PairSwapMatchedLen, "must match once"},
+		// Provide
+		{PairProvideRawLogStr, nil, CreatePairCommonRulesFinder, 1, PairProvideMatchedLen, "must match once"},
+		// Withdraw
+		{PairWithdrawRawLogStr, nil, CreatePairCommonRulesFinder, 1, PairWithdrawMatchedLen, "must match once"},
+	}
+
+	for idx, tc := range tcs {
+		errMsg := fmt.Sprintf("idx(%d): %s", idx, tc.errMsg)
+		assert := assert.New(t)
+
+		setUp(tc.rawLogStr, tc.pairs, tc.finderFunc)
+		matchedResults := logFinder.FindFromLogs(eventLogs)
+		assert.Len(matchedResults, tc.expectedResultLen, errMsg)
+		if tc.expectedResultLen > 0 {
+			assert.Len(matchedResults[0], tc.matchedLen, "must return all matched value")
```

Comment:
The test fails in my local at this line. Is there any fix from the following changes?

Reply:
Yes, I divided PR for readability.